src/DominantJuice.sol

```diff
@@ -0,0 +1,399 @@
+
+    /// Main Campaign Parameters
+    struct Campaign {
+        uint256 projectId;
+        uint256 cycleTarget;
+        uint256 cycleExpiryDate;
+        uint256 minimumPledgeAmount;
+        uint256 totalRefundBonus;
+    }
+
+    struct FundingStatus {
+        uint256 totalPledged;
+        uint256 percentOfGoal;
+        bool isTargetMet;
+        bool hasCycleExpired;
+        bool hasCreatorWithdrawnAllFunds;
+    }
+
+    struct Pledgers {
+        uint256 totalAmountPledged;
+        UD60x18 totalPledgeWeight;
+        mapping(address => UD60x18) pledgerWeight;
+        mapping(address => bool) hasBeenRefunded;
+    }
+
+    /// State Variables
+    Campaign private campaign;
+    Pledgers private pledgers;
+    UD60x18 private immutable rateOfDecay = ud(0.99e18);
+
+    // Events for transparency to pledgers
+    event RefundBonusDeposited(address, uint256 indexed);
+    event PledgeMade(address, uint256);
+    event CycleHasClosed(bool indexed, bool indexed);
+    event CycleRefundBonusWithdrawal(address indexed, uint256 indexed);
+    event CreatorWithdrawal(address, uint256);
+
+    error BonusAlreadyDeposited(uint256 bonusAmount);
+    error RefundBonusNotDeposited();
+    error AmountIsBelowMinimumPledge(uint256 minAmount);
+    error InvalidPaymentEvent(address caller, uint256 projectId, uint256 value);
+    error CycleHasExpired();
+    error IncorrectProjectID();
+    error FunctionHasAlreadyBeenCalled();
+    error MustBePledger();
+    error CycleHasNotEndedYet(uint256 endTimestamp);
+    error NoRefundsForSuccessfulCycle();
+    error AlreadyWithdrawnRefund();
+    error InvalidRedemptionEvent(address caller, uint256 projectId, uint256 value);
+    error InsufficientFunds();
+
+    // Make sure the caller is a terminal of the project, and that the call is being made
+    // on behalf of an interaction with the correct project.
+    modifier terminalCheck() {
+        require(msg.value == 0, "Pledges should be made through JB website.");
+        require(
+            directory.isTerminalOf(campaign.projectId, IJBPaymentTerminal(msg.sender)),
+            "Caller must be a JB Payment Terminal."
+        );
+        _;
+    }
+
+    modifier redeemCycleCheck() {
+        // This is an insurance check, since project should have redemptions paused during 1st cycle.
+        if (!hasCycleExpired()) revert CycleHasNotEndedYet(campaign.cycleExpiryDate);
+        if (isTargetMet()) revert NoRefundsForSuccessfulCycle();
+        _;
+    }
+
+    /// @param _projectId Obtained via Juicebox after project creation.
+    constructor(
+        address payable _campaignManager,
+        uint256 _projectId,
+        uint256 _cycleTarget,
+        uint256 _minimumPledgeAmount,
+        IJBController3_1 _controller,
+        IJBSingleTokenPaymentTerminalStore3_1_1 _paymentTerminalStore
+    ) {
+        // Assign roles
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(CAMPAIGN_MANAGER_ROLE, _campaignManager);
```

Comment:
I don't like to grant other roles during construction. The external grantRole function should be used by the admin after construction.

Reply:
Okay, I’ll remove and add in the campaign workflow. Will refactor the tests as well to reflect this change.